tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py

```diff
@@ -281,12 +287,77 @@ def _zeros_slot(var, slot_name, op_name):
       named_slots[optimizer._var_key(var)] = new_slot_variable
     return named_slots[optimizer._var_key(var)]
 
+  def apply_gradients(grads_and_vars, global_step=None, name=None):
+    """Apply gradients to variables.
+    Args:
+      grads_and_vars: List of (gradient, variable) pairs as returned by
+        compute_gradients().
+      global_step: Optional Variable to increment by one after the
+        variables have been updated.
+      name: Optional name for the returned operation.  Default to the
+        name passed to the Optimizer constructor.
+
+    Returns:
+      train_op: apply gradients op to be executed by each replica.
+
+    Raises:
+      ValueError: If the grads_and_vars is empty.
+      ValueError: If global step is not provided, the staleness cannot be
+        checked.
+    """
+    try:
+      import horovod.tensorflow as hvd
+    except ImportError:
+      raise ValueError(
+          "Please install Horovod first if you want to use distributed synchronous training ased on Horovod"
+      )
+    if not grads_and_vars:
+      raise ValueError("Must supply at least one variable")
+
+    if global_step is None:
+      raise ValueError("Global step is required to check staleness")
+
+    trainable_grad_and_vars = []
+    aggregated_grad = []
+    var_list = []
+
+    with backend.name_scope(name or self._name):
+      for grad, var in grads_and_vars:
+        if isinstance(var, de.TrainableWrapper):
+          trainable_grad_and_vars.append((grad, var))
+          continue
+        var_list.append(var)
+        with ops.device(var.device):
+          # Dense gradients.
+          if grad is None:
+            aggregated_grad.append(None)  # pass-through.
+            continue
+          elif isinstance(grad, ops.Tensor):
+            aggregated_grad.append(hvd.allreduce(grad, op=hvd.Sum))
```

Comment:
And just for discussion, I noticed that Horovod implements lots of features and optimizations on its DistributedOptimizer like tensor fusion, grouped allreduce, adasum, gradient compression, etc.., in the latest-nth versions. Use the hvd.allreduce API may not reuse these features.

Reply:
Yeah, For `tensor fusion`, it's set by environment variable, and it's turn on by default, for most recommend situation, I think there is no need to care about `grouped allreduce`  or `gradient compression`. But I think we could find better way to open `reduction operation` to the user. If we directly let user specify `hvd.Sum` or `hvd.adamsum`, we should let them import `horovod` before `apply_gradients`, or we could let user specify a string the reduction method like `sum`, we map this method to the corresponding `hvd op`, but if  horovod add some reduction op, we should change the code to be compatible with newest horovod version.